lib/logstash/inputs/file.rb

```diff
@@ -160,176 +165,241 @@ class LogStash::Inputs::File < LogStash::Inputs::Base
-      @path, @input = path, input
-    end
-
-    def timed_out
-      input.codec.evict(path)
-    end
-
-    def accept(data)
-      # and push transient data filled dup listener downstream
-      input.log_line_received(path, data)
-      input.codec.accept(dup_adding_state(data))
-    end
-
-    def process_event(event)
-      event.set("[@metadata][path]", path)
-      event.set("path", path) if !event.include?("path")
-      input.post_process_this(event)
-    end
-
-    def add_state(data)
-      @data = data
-      self
-    end
-
-    private
-
-    # duplicate and add state for downstream
-    def dup_adding_state(line)
-      self.class.new(path, input).add_state(line)
-    end
-  end
-
-  class FlushableListener < ListenerTail
-    attr_writer :path
-  end
-
   def listener_for(path)
     # path is the identity
-    ListenerTail.new(path, self)
+    FileListener.new(path, self)
   end
 
-  def begin_tailing
+  def start_processing
     # if the pipeline restarts this input,
     # make sure previous files are closed
     stop
-    # use observer listener api
-    @tail = FileWatch::Tail.new_observing(@tail_config)
-    @tail.logger = @logger
-    @path.each { |path| @tail.tail(path) }
+    @watcher = @watcher_class.new(@filewatch_config)
+    @path.each { |path| @watcher.watch_this(path) }
   end
 
   def run(queue)
-    begin_tailing
+    start_processing
     @queue = queue
-    @tail.subscribe(self)
+    @watcher.subscribe(self) # halts here until quit is called
     exit_flush
   end # def run
 
   def post_process_this(event)
     event.set("[@metadata][host]", @host)
-    event.set("host", @host) if !event.include?("host")
+    event.set("host", @host) unless event.include?("host")
     decorate(event)
     @queue << event
   end
 
+  def handle_deletable_path(path)
+    return if tail_mode?
+    return if @completed_file_handlers.empty?
+    @completed_file_handlers.each { |handler| handler.handle(path) }
+  end
+
   def log_line_received(path, line)
-    return if !@logger.debug?
+    return unless @logger.debug?
     @logger.debug("Received line", :path => path, :text => line)
   end
 
   def stop
     # in filewatch >= 0.6.7, quit will closes and forget all files
     # but it will write their last read positions to since_db
     # beforehand
-    if @tail
+    if @watcher
       @codec.close
-      @tail.quit
+      @watcher.quit
     end
   end
 
   private
 
+  def build_sincedb_base_from_settings(settings)
+    logstash_data_path = settings.get_value("path.data")
+    Pathname.new(logstash_data_path).join("plugins", "inputs", "file").tap do |path|
+      # Ensure that the filepath exists before writing, since it's deeply nested.
+      path.mkpath
+    end
+  end
+
+  def build_sincedb_base_from_env
+    # This section is going to be deprecated eventually, as path.data will be
+    # the default, not an environment variable (SINCEDB_DIR or HOME)
+    if ENV["SINCEDB_DIR"].nil? && ENV["HOME"].nil?
```

Comment:
as previoustly noted, I think we need to verify all ENV usages here for `HOME`, `LS_HOME`, `LOGSTASH_HOME`, `SINCEDB_DIR`, `SINCEDB_PATH` ...